# Purpose:

For this analysis, we are going to use 2008 alirlines dataset from the Data expository 09 of ASA website: http://stat-computing.org/dataexpo/2009/the-data.html. We have extracted this dataset directly from this website to our current working directory. In this analysis, we are going demonstrate how to deal with missing values in dataset.

# Step One:

We need to import some necessary python libraries for this analysis with the following python commands.

In [1]:
import os
import pandas as pd
import numpy as np

With the following python command, we can print the current working directory.

In [2]:
print(os.getcwd())

C:\Users\Owner\Desktop\New airlines


### Import dataset:

To import the Airlines_2008.csv file from the current working directory to this notebook, we use the following python command. 

In [3]:
Airlines = pd.read_csv('Airlines_2008.csv')

### Structure of the dataset:

To see the structure of this dataset, we use the following python command.

In [4]:
Airlines.info(memory_usage = 'deep', null_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7009728 entries, 0 to 7009727
Data columns (total 29 columns):
Year                 7009728 non-null int64
Month                7009728 non-null int64
DayofMonth           7009728 non-null int64
DayOfWeek            7009728 non-null int64
DepTime              6873482 non-null float64
CRSDepTime           7009728 non-null int64
ArrTime              6858079 non-null float64
CRSArrTime           7009728 non-null int64
UniqueCarrier        7009728 non-null object
FlightNum            7009728 non-null int64
TailNum              6926363 non-null object
ActualElapsedTime    6855029 non-null float64
CRSElapsedTime       7008884 non-null float64
AirTime              6855029 non-null float64
ArrDelay             6855029 non-null float64
DepDelay             6873482 non-null float64
Origin               7009728 non-null object
Dest                 7009728 non-null object
Distance             7009728 non-null int64
TaxiIn               6858079 non-

The Airlines dataset is a pandas DataFrame. This DataFrame contains 7,009,728 rows from 0 to 7,009,727 and 29 columns. 14 of these columns are float type data, 10 of them are integer type data, and 5 of them are object type data. Since some values in the second column above are less than 7,009,728, there are missing values in this dataset. To see these missing values, we use the following python command. 

In [5]:
Airlines.isnull().sum(axis = 0)

Year                       0
Month                      0
DayofMonth                 0
DayOfWeek                  0
DepTime               136246
CRSDepTime                 0
ArrTime               151649
CRSArrTime                 0
UniqueCarrier              0
FlightNum                  0
TailNum                83365
ActualElapsedTime     154699
CRSElapsedTime           844
AirTime               154699
ArrDelay              154699
DepDelay              136246
Origin                     0
Dest                       0
Distance                   0
TaxiIn                151649
TaxiOut               137058
Cancelled                  0
CancellationCode     6872294
Diverted                   0
CarrierDelay         5484993
WeatherDelay         5484993
NASDelay             5484993
SecurityDelay        5484993
LateAircraftDelay    5484993
dtype: int64

The values of the second column of the above table show the number of missing values; 0 means there is no missing value in that column, otherwise there are that many missing values in that column.  

# Step Two:

### How to deal with missing values?

 - The easiest way to deal with missing values is removing all the rows that contain any missing cells. The downside of this is that we might be losing some vital information. Therefore, we also consider placing values in such cells. 

**For numerical missing values:**

 - For numerical missing cells, we can replace them with a mean or a median of the non-missing cells of that column.

**For categorical missing values:**

 - For categorical missing cells, we can replace them with a unknown/NA or 'Missing Value" or some kind of strange value like "99999" or any value that is more make sense for the particular context. 

### Removing all the rows that contains any missing cell values.

In [6]:
Airlines1 = Airlines.dropna(how = 'any', axis = 0)
Airlines1.shape

(0, 29)

The result of the above python command indicates that every row of the Airlines dataset contains at least one missing cell value; that is because most of the cells of the column CancellationCode are missing. Let us remove the CancellationCode column from the Airlines dataset and recalculate the missing values.          

In [7]:
Airlines2 = Airlines.iloc[:, [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21, 23,24,25,26,27,28]]
Airlines2.shape

(7009728, 28)

After removing the CancellationCode column, this dataset is left with 28 columns instead of 29 columns. Let us recheck how many rows are incomplete.

In [8]:
Airlines2.isnull().sum(axis = 0)

Year                       0
Month                      0
DayofMonth                 0
DayOfWeek                  0
DepTime               136246
CRSDepTime                 0
ArrTime               151649
CRSArrTime                 0
UniqueCarrier              0
FlightNum                  0
TailNum                83365
ActualElapsedTime     154699
CRSElapsedTime           844
AirTime               154699
ArrDelay              154699
DepDelay              136246
Origin                     0
Dest                       0
Distance                   0
TaxiIn                151649
TaxiOut               137058
Cancelled                  0
Diverted                   0
CarrierDelay         5484993
WeatherDelay         5484993
NASDelay             5484993
SecurityDelay        5484993
LateAircraftDelay    5484993
dtype: int64

### Remove rows with any colums with missing cell values:

Removing all the rows from the modified Airlines dataset that contain any missing cell values.

In [9]:
Airlines3 = Airlines2.dropna(how = 'any', axis = 0)
Airlines3.shape

(1524733, 28)

Now number of rows are reduced to 1,524,733 from 7,009,728; that is because the removed rows contain missing values.

# Step Three:

Imputation of the missing values. Let us make a copy of the Airlines dataset and give it the name Air2008.

In [10]:
Air2008 = Airlines

### Let us check again the missing values in the Air2008:

To see if Air2008 contains any missing values, we use the following python command.

In [11]:
Air2008.isnull().sum(axis = 0)

Year                       0
Month                      0
DayofMonth                 0
DayOfWeek                  0
DepTime               136246
CRSDepTime                 0
ArrTime               151649
CRSArrTime                 0
UniqueCarrier              0
FlightNum                  0
TailNum                83365
ActualElapsedTime     154699
CRSElapsedTime           844
AirTime               154699
ArrDelay              154699
DepDelay              136246
Origin                     0
Dest                       0
Distance                   0
TaxiIn                151649
TaxiOut               137058
Cancelled                  0
CancellationCode     6872294
Diverted                   0
CarrierDelay         5484993
WeatherDelay         5484993
NASDelay             5484993
SecurityDelay        5484993
LateAircraftDelay    5484993
dtype: int64

By looking at the above table, it is clear that this dataset contains missing cell values. We can impute values to some of these missing cells.  

### Imputation with missing values: 

If the distribution is symmetric, then the mean imputation is better. If the distribution is skewed then the median imputation is better. This is because, the mean is affected by the extreme values and the median is not affected by the extreme values.

In [12]:
% matplotlib inline

import seaborn as sns

# Mean Imputation:

The variable DepTime, departure time, contains 136,246 missing cell values and it is a float type data.

In [13]:
Air2008[Air2008.DepTime.isnull()].shape

(136246, 29)

6,873,482 rows does not contain any missing values.

In [14]:
Air2008[Air2008.DepTime.notnull()].shape

(6873482, 29)

### To fill missing cell values with a mean:

We are going to fill all the NaN values with a mean of non missing cells of the departure time, DepTime, column with the following python command. 

In [15]:
Air2008['DepTime'].fillna(value = np.nanmean(Air2008.DepTime), inplace = True)

With the following python command, we can see the shape of the departure time column.

In [16]:
Air2008.DepTime.shape

(7009728,)

The shape of the departure time column above shows that its all the missing cell values are filled.

# Median Imputation:

The variable ArrTime, arrival time, contains 151,649 missing cell values and it is a float type data.

In [17]:
Air2008[Air2008.ArrTime.isnull()].shape

(151649, 29)

6,858,079 cells of the Arrival time variable are non-missing values.

In [18]:
Air2008[Air2008.ArrTime.notnull()].shape

(6858079, 29)

### To fill with a median:

We are going to fill all the NaN values with a median of non missing cells of ArrTime column with the following python
command. 

In [19]:
Air2008['ArrTime'].fillna(value = np.nanmedian(Air2008.ArrTime), inplace = True)

In [20]:
Air2008.ArrTime.shape

(7009728,)

The shape of the Arrival time variable shows all the missing cell values are filled.

# Object Imputation:

The variable CancellationCode contains 6,872,294 missing cell values and it is a object type data.

In [21]:
Air2008[Air2008.CancellationCode.isnull()].shape

(6872294, 29)

137,434 rows do not have any missing values.

In [22]:
Air2008[Air2008.CancellationCode.notnull()].shape

(137434, 29)

To see the number of unique values in a variable, CancellationCode, we use following python command.

In [23]:
Air2008.CancellationCode.unique()

array([nan, 'A', 'C', 'B', 'D'], dtype=object)

To see the each of the unique values in a dataset, CancellationCode, we use the following python command, including NaN values.

In [24]:
Air2008.CancellationCode.value_counts(dropna = False)

NaN    6872294
B        54904
A        54330
C        28188
D           12
Name: CancellationCode, dtype: int64

### To fill with a object, Missing:

We are going to replace all the NaN values of the ActualElapsedTime column with a 'Missing' by using following python command. 

In [25]:
Air2008['CancellationCode'].fillna(value = 'Missing', inplace = True)

In [26]:
Air2008.CancellationCode.shape

(7009728,)

The shape of the CancellationCode variable above shows that all the missing cell values are filled. 

We investigate the number of missing cell values of the columns in a dataset, Air2008, with the following python command.

In [27]:
Air2008.isnull().sum(axis = 0)

Year                       0
Month                      0
DayofMonth                 0
DayOfWeek                  0
DepTime                    0
CRSDepTime                 0
ArrTime                    0
CRSArrTime                 0
UniqueCarrier              0
FlightNum                  0
TailNum                83365
ActualElapsedTime     154699
CRSElapsedTime           844
AirTime               154699
ArrDelay              154699
DepDelay              136246
Origin                     0
Dest                       0
Distance                   0
TaxiIn                151649
TaxiOut               137058
Cancelled                  0
CancellationCode           0
Diverted                   0
CarrierDelay         5484993
WeatherDelay         5484993
NASDelay             5484993
SecurityDelay        5484993
LateAircraftDelay    5484993
dtype: int64

Still many of the columns of this dataset contains missing cell values. 

### Droping intersection of the columns:

Drop a row only if both of these columns' cells contain NaNs.

In [28]:
Air2008.dropna(subset = ['AirTime', 'ArrDelay'], how = 'all', inplace = True)

To see above have worked, we use following python command.

In [29]:
Air2008.isnull().sum(axis = 0)

Year                       0
Month                      0
DayofMonth                 0
DayOfWeek                  0
DepTime                    0
CRSDepTime                 0
ArrTime                    0
CRSArrTime                 0
UniqueCarrier              0
FlightNum                  0
TailNum                    5
ActualElapsedTime          0
CRSElapsedTime             0
AirTime                    0
ArrDelay                   0
DepDelay                   0
Origin                     0
Dest                       0
Distance                   0
TaxiIn                     0
TaxiOut                    0
Cancelled                  0
CancellationCode           0
Diverted                   0
CarrierDelay         5330294
WeatherDelay         5330294
NASDelay             5330294
SecurityDelay        5330294
LateAircraftDelay    5330294
dtype: int64

### Dropping a union of the columns:

Drop a row if either one of these columns contain NaN value. 

In [30]:
Air2008.dropna(subset = ['CarrierDelay', 'WeatherDelay'], how = 'any', inplace = True)

To see above worked, we use python command.

In [31]:
Air2008.isnull().sum(axis = 0)

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              2
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

To see above two cells missing values of variable TailNum, we use following python command.

In [32]:
Air2008[Air2008.TailNum.isnull()]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1643793  2008      3          18          2   1716.0        1655   2036.0   
2556354  2008      5          31          6   1720.0        1244   1920.0   

         CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  \
1643793        1947            9E       3760        ...           11.0   
2556354        1427            OH       5396        ...            8.0   

         TaxiOut  Cancelled  CancellationCode  Diverted  CarrierDelay  \
1643793     29.0          0           Missing         0           0.0   
2556354     40.0          0           Missing         0           0.0   

        WeatherDelay NASDelay  SecurityDelay  LateAircraftDelay  
1643793          0.0     28.0            0.0               21.0  
2556354        276.0     17.0            0.0                0.0  

[2 rows x 29 columns]

To remove above two rows from Air2008 dataset, we use the following python command and give it a name Air2008_copy.

In [33]:
Air2008_copy = Air2008.iloc[((Air2008.index != 1643793) & (Air2008.index != 2556354)), :]

To see the structure of the dataset, Air2008_copy, we use the following python command. 

In [34]:
Air2008_copy.info(memory_usage = 'deep', null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1524733 entries, 4 to 7009721
Data columns (total 29 columns):
Year                 1524733 non-null int64
Month                1524733 non-null int64
DayofMonth           1524733 non-null int64
DayOfWeek            1524733 non-null int64
DepTime              1524733 non-null float64
CRSDepTime           1524733 non-null int64
ArrTime              1524733 non-null float64
CRSArrTime           1524733 non-null int64
UniqueCarrier        1524733 non-null object
FlightNum            1524733 non-null int64
TailNum              1524733 non-null object
ActualElapsedTime    1524733 non-null float64
CRSElapsedTime       1524733 non-null float64
AirTime              1524733 non-null float64
ArrDelay             1524733 non-null float64
DepDelay             1524733 non-null float64
Origin               1524733 non-null object
Dest                 1524733 non-null object
Distance             1524733 non-null int64
TaxiIn               1524733 non-

### Exporting Dataset:

To export Air2008_copy to current working directory.

In [35]:
Air2008_copy.to_csv('Air2008_copy.csv', index=False)
print('Done')

Done


In [36]:
Air2008_copy.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
4   2008      1           3          4   1829.0        1755   1959.0   
6   2008      1           3          4   1937.0        1830   2037.0   
11  2008      1           3          4   1644.0        1510   1845.0   
16  2008      1           3          4   1452.0        1425   1640.0   
18  2008      1           3          4   1323.0        1255   1526.0   

    CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  TaxiOut  \
4         1925            WN       3920        ...            3.0     10.0   
6         1940            WN        509        ...            3.0      7.0   
11        1725            WN       1333        ...            6.0      8.0   
16        1625            WN        675        ...            7.0      8.0   
18        1510            WN          4        ...            4.0      9.0   

    Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
4           0           Missing         0           2.0          0.0      0.0   
6           0           Missing         0          10.0          0.0      0.0   
11          0           Missing         0           8.0          0.0      0.0   
16          0           Missing         0           3.0          0.0      0.0   
18          0           Missing         0           0.0          0.0      0.0   

    SecurityDelay  LateAircraftDelay  
4             0.0               32.0  
6             0.0               47.0  
11            0.0               72.0  
16            0.0               12.0  
18            0.0               16.0  

[5 rows x 29 columns]